File: A06 Nodes Related/002 Binary Tree/105. Construct Binary Tree from Preorder and Inorder Traversal/Lc105.py
###########
Time Complexity:
Space Complexity:


#############

  def buildTree(self, preorder: List[int], inorder: List[int]) -> TreeNode:
        
        if not preorder or not inorder:
            return None

        if len(preorder) == 1:
            return TreeNode(preorder[0])
        
        root = TreeNode(preorder[0])
        root_index = inorder.index(preorder[0])
       
        root.left = self.buildTree(preorder[1:root_index + 1],inorder[:root_index])
        root.right = self.buildTree(preorder[root_index + 1:],inorder[root_index + 1:])
        
        return root

#################

def buildTree(self, preorder, inorder):
    if inorder:
        ind = inorder.index(preorder.pop(0))
        root = TreeNode(inorder[ind])
        root.left = self.buildTree(preorder, inorder[0:ind])
        root.right = self.buildTree(preorder, inorder[ind+1:])
        return root

########
  def buildTree(self, preorder, inorder):
  
        if len(inorder) == 0:
            return None

        if len(inorder) == 1:
            return TreeNode(preorder[0])
        
        rval = preorder[0]
        root = TreeNode(rval)
        
        inorder_rval_index = inorder.index(rval)
        left_inorder = inorder[:inorder_rval_index]
        right_inorder = inorder[inorder_rval_index+1:]
        left_preorder = preorder[1: len(left_inorder) + 1]
        right_preorder = preorder[1 + len(left_preorder):]
        
        root.left = self.buildTree(left_preorder, left_inorder)
        root.right = self.buildTree(right_preorder, right_inorder)
        
        return root

File: A06 Nodes Related/002 Binary Tree/105. Construct Binary Tree from Preorder and Inorder Traversal/LC105.java
/*
Time Complexity:O(n)
Space Complexity:  O(n)

*/ 


/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public TreeNode buildTree(int[] preorder, int[] inorder) {
        HashMap<Integer, Integer> map = new HashMap<>(); 
        for (int i = 0; i < inorder.length; i ++){
            map.put(inorder[i], i);
        }
        
        return helper(preorder, inorder, 0, 0, inorder.length -1, map);
    }
    
    private TreeNode helper(int[] preorder, int[] inorder, int preStart, 
                            int inStart, int inEnd, HashMap<Integer, Integer> map){
        //base case
        if (preStart >= preorder.length || inStart > inEnd) return null; 
        
        TreeNode root = new TreeNode(preorder[preStart]); 
        int rootIndex = map.get(preorder[preStart]);
        
        root.left = helper(preorder, inorder, preStart +1, inStart, rootIndex -1, map); 
        root.right = helper(preorder, inorder, preStart + rootIndex - inStart +1, rootIndex +1, inEnd, map);
        
        return root; 
    }
}/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public TreeNode buildTree(int[] preorder, int[] inorder) {
        HashMap<Integer, Integer> map = new HashMap<>(); 
        for (int i = 0; i < inorder.length; i ++){
            map.put(inorder[i], i);
        }
        
        return helper(preorder, inorder, 0, 0, inorder.length -1, map);
    }
    
    private TreeNode helper(int[] preorder, int[] inorder, int preStart, 
                            int inStart, int inEnd, HashMap<Integer, Integer> map){
        //base case
        if (preStart >= preorder.length || inStart > inEnd) return null; 
        
        TreeNode root = new TreeNode(preorder[preStart]); 
        int rootIndex = map.get(preorder[preStart]);
        
        root.left = helper(preorder, inorder, preStart +1, inStart, rootIndex -1, map); 
        root.right = helper(preorder, inorder, preStart + rootIndex - inStart +1, rootIndex +1, inEnd, map);
        
        return root; 
    }
}